grails-app/services/org/pih/warehouse/invoice/PrepaymentInvoiceService.groovy

```diff
@@ -23,47 +57,222 @@ class PrepaymentInvoiceService {
+    }
+
+    /**
+     * Updates invoice item's quantity and amount and finds related inverse item and updates quantity and amount
+     * */
+    void updateInvoiceItemQuantity(String itemId, Integer quantity) {
+        InvoiceItem invoiceItem = InvoiceItem.get(itemId)
+        if (quantity > invoiceItem.quantity) {
+            throw new IllegalArgumentException("Cannot update quantity invoiced to higher value")
         }
+        if (quantity <= 0) {
+            throw new IllegalArgumentException("Quantity invoiced needs to be higher than 0")
+        }
+        // I think this entire update is only for shipment items (but will double check that)
+        if (!invoiceItem?.shipmentItem) {
+            throw new IllegalArgumentException("Invoice item is missing shipment item. Cannot edit quantity.")
+        }
+        // update invoice item's quantity (and adjust amount)
+        invoiceItem.quantity = quantity
+        invoiceItem.amount = quantity * invoiceItem.unitPrice
+        // update inverse item's quantity (and adjust amount)
+        OrderItem orderItem = invoiceItem?.shipmentItem?.orderItems?.find { it }
+        BigDecimal prepaymentPercent = (orderItem.order.paymentTerm?.prepaymentPercent ?: Constants.DEFAULT_PAYMENT_PERCENT) / 100
+        InvoiceItem prepaymentItem = orderItem.invoiceItems.find { it.isPrepaymentInvoice }
+        InvoiceItem inverseItem = findInverseItem(invoiceItem)
+        inverseItem.quantity = quantity
+        inverseItem.amount = quantity * prepaymentItem.unitPrice * prepaymentPercent * (-1)
+    }
 
-        order.activeOrderAdjustments.each { OrderAdjustment orderAdjustment ->
-            InvoiceItem invoiceItem = invoiceService.createFromOrderAdjustment(orderAdjustment)
-            invoice.addToInvoiceItems(invoiceItem)
+    /**
+     * Removes invoice item for the itemId and finds related inverse item and removes it too
+     * */
+    void removeInvoiceItem(String itemId) {
+        InvoiceItem invoiceItem = InvoiceItem.get(itemId)
+        if (!invoiceItem) {
+            throw new IllegalArgumentException("Missing invoice item to delete")
+        }
+        if (invoiceItem.invoiceItemType != InvoiceItemType.REGULAR) {
+            throw new IllegalArgumentException("Cannot delete items with type other than REGULAR")
         }
 
-        return invoice.save()
+        InvoiceItem inverseItem = findInverseItem(invoiceItem)
+        if (inverseItem) {
+            deleteInvoiceItem(inverseItem)
+        }
+        deleteInvoiceItem(invoiceItem)
+    }
+
+    /**
+     * Clean up related objects and delete item
+     * */
+    private void deleteInvoiceItem(InvoiceItem invoiceItem) {
+        invoiceItem.orderAdjustments?.each { OrderAdjustment oa -> oa.removeFromInvoiceItems(invoiceItem) }
+        invoiceItem.orderItems?.each { OrderItem oi -> oi.removeFromInvoiceItems(invoiceItem) }
+        invoiceItem.shipmentItems?.each { ShipmentItem si -> si.removeFromInvoiceItems(invoiceItem) }
```

Comment:
is there no way in GORM to set this to be automatic? If an item is deleted, remove it from all relationships? It's weird to me that this needs to be manual

Reply:
@EWaterman These are many to many relations, and there is the case that cascade is for one to many only and when we still have invoice item instance on the related object (order item, order adjustment, or shipment item), then there would be thrown
```
org.springframework.dao.InvalidDataAccessApiUsageException: deleted object would be re-saved by cascade (remove deleted object from associations)
```
See:
https://spring.io/blog/2010/07/02/gorm-gotchas-part-2#deletes-do-not-cascade-period
https://grails.github.io/legacy-grails-doc/3.2.0/ref/Database%20Mapping/cascade.html